blog/_posts/2023-05-26-scala-3.3.0-released.md

```diff
@@ -0,0 +1,162 @@
+
+```Scala
+def canFail(input: String) = Try:
+  someComputation(input).flatMap: res =>
+    val partial = moreComputation(res)
+    andEvenMore(partial)
+
+class TestSuite extends munit.FunSuite:
+  test("the job doesn't fail"):
+    val job = canFail("some data")
+    assert(job.isSuccess)
+```
+
+### `boundary` and `break`
+
+Two new methods were added to the standard library in the place: `boundary` and `brake`. They are safer and more expressive replacements for recently deprecated non-local returns.
+
+`break` allows for a type-safe early escape from anywhere inside the block delimited by `boundary` to its end, returning the passed value from the entire block.
+
+```Scala
+import util.boundary, boundary.break
+
+def sumOfRoots(number: List[Double]): Option[Double] = boundary {
+  val roots = numbers.map { n =>
+    println(s" * calculating square root for $n*")
+    if n >= 0 then Math.sqrt(n) else break(None)
+  }
+  Some(roots.sum)
+}
+```
+
+When you run the above method, you will notice that it returns None when any of the input elements is negative. Moreover, thanks to `boundary`/`break`, you can see from the console output that it stops the entire process after encountering the first negative element.
+
+`break` can jump out of the`boundary` in the function deeper on the stack. To make it safe, only functions with a matching `Label` in their using clauses can break. This can be used to create isolated parts of an application with streamlined error handling.
+
+Library authors can abstract over `break` and `Label`, creating a nice API for error handling or dealing with uncertain data. We can define the following micro-library as a simplified example:
+
+```Scala
+import util.boundary, boundary.{ Label, break }
+
+type Uncertain[T] = Label[Left[String, Nothing]] ?=> T
+
+def fail(error: String): Uncertain[Nothing] = break(Left(error))
+```
+
+Then it can be used in the following way to create a fragment of a larger application with isolated error handling:
+
+```Scala
+def getUser(id: String): Uncertain[User] =
+  val user = findUser(id)
+  if verifyUser(user) then user else fail("Incorrect user")
+
+def getAddress(user: User): Uncertain[Address] =
+  findAddress(user) match
+    case Some(address) if verifyAddress(address) => address
+    case Some(_) => fail("Incorrect address")
+    case None => fail("Missing address")
+
+def getShippingData(id: String): Either[String, (User, Address)] =
+  boundary {
+    val user = getUser(id)
+    val address = getAddress(user)
+    Right((user, address))
+  }
+```
+
+### The new default implementation of lazy vals
+
+Last but not least, we have changed the default implementation of lazy vals. The new implementation has better performance and is safer under parallel access. This can lead to improvements in the effect systems performance.
```

Comment:
```suggestion
Last but not least, we have changed the default implementation of lazy vals. The new implementation has better performance and is safer under parallel access. This can lead to performance improvements in the effect systems.
```

(but which effect systems do we talk about?)

Reply:
just "effect systems", not "the effect systems"

Reply:
This may result in improvements in the performance of effect systems.